### DISCUSSION TITLE
Conan: single_include missing in package

### DISCUSSION THREAD

**Discussion by nic-godz** [nlohmann/json#3483] [Q&A]
As far as I can see the **single_include**  directory with the json.hpp with minimized includes are missing in the Conan package:

`/data/.conan/data/nlohmann_json/3.10.5/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include/nlohmann/json.hpp
/data/.conan/data/nlohmann_json/3.10.5/_/_/source/source_subfolder/include/nlohmann/json.hpp
/data/.conan/data/nlohmann_json/3.10.5/_/_/source/source_subfolder/single_include/nlohmann/json.hpp`

Anyone else using Conan and missing the "full" json.hpp file? I guess it this one needs as well:
`/data/.conan/data/nlohmann_json/3.10.5/_/_/package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/single_include/nlohmann/json.hpp`

Or preferably only the single_include as for cmake installation the "multi" version needs extra options for installation

**Comment by falbrechtskirchinger** [comment]
The Conan recipes are maintained by the Conan project. If something is missing, asking them directly is the better choice.

Also, I see `single_include/nlohmann/json.hpp` in your output, so what exactly are you missing?

Edit: I guess that is the package source and `package/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/include/nlohmann/json.hpp` is the "installed" header? Are you sure the directory is not just renamed? If it's the amalgamated header it should be around 20k lines long.

**Reply by nic-godz** [reply]
Yes, single_include is only included in the source folder and missing in the package folder. Yeah, I first thought it was just a matter of renaming but the json.hpp in the package folder is just 4.7k lines long.

I'm pretty far from being a Conan expert but have played around with som Conan recipes. I have used rules like:
`exports_sources = "include/*"`

So I guess rules for single_include folder is needed as well in order for it to end up in the package?

**Reply by falbrechtskirchinger** [reply]
More importantly: why do you insist on using the amalgamated header when you're using Conan? The amalgamated header exists for easy integration. You are solving the integration issue by using Conan.
To use the library, just include `<nlohmann/json.hpp>`. Is that not working for you?

**Reply by nic-godz** [reply]
I'm moving over from having some 3rd-party headers in the source tree to using Conan. Got some build issues going form the 20k lines "easy integration" json.hpp to the dependency heavy one. Nothing that couldn't be solved.

But as **cmake/make install** installs(default) only the 20k lines json.hpp file I thought that was the intention with Conan as well. If so, the recipe needs to be changed, see below.

But if the intention with **cmake/make install** is to install the 20k lines file and Conan to do the opposite and package the 4.7k lines file I'm fine with that. Not a big deal.

**Reply by falbrechtskirchinger** [reply]
> But as **cmake/make install** installs(default) only the 20k lines json.hpp file I thought that was the intention with Conan as well. If so, the recipe needs to be changed, see below.

Default being the operative word. It installs one or the other depending on the `JSON_MultipleHeaders` setting:
```cmake
if (JSON_MultipleHeaders)
    set(NLOHMANN_JSON_INCLUDE_BUILD_DIR "${PROJECT_SOURCE_DIR}/include/")
    message(STATUS "Using the multi-header code from ${NLOHMANN_JSON_INCLUDE_BUILD_DIR}")
else()
    set(NLOHMANN_JSON_INCLUDE_BUILD_DIR "${PROJECT_SOURCE_DIR}/single_include/")
    message(STATUS "Using the single-header code from ${NLOHMANN_JSON_INCLUDE_BUILD_DIR}")
endif()
```

My Linux distribution (Gentoo) installs the multi-header version. Ubuntu seems to install the amalgamated header according to the [file list](https://packages.ubuntu.com/bionic/all/nlohmann-json-dev/filelist):
```
/usr/include/json.hpp
/usr/include/nlohmann/json.hpp
/usr/lib/cmake/nlohmann_jsonConfig.cmake
```

**Comment by nlohmann** [comment]
Though we are not maintaining the Conan recipe ourselves, we would still be happy about feedback on the documentation on this: https://json.nlohmann.me/integration/package_managers/#conan - is there anything missing?

**Reply by nic-godz** [reply]
I see, I checked the Conan recipe and I'm by far a Conan expert but think it should look like:
`    def package(self):
        self.copy(pattern="LICENSE*", src=self._source_subfolder, dst="licenses")
        self.copy("*", src=os.path.join(self._source_subfolder, "single_include"), dst="include")`

That will hopefully export the **single_include** directory instead of **include** but still package it as **include**. If **include** directory(with multi- json.hpp) is need another **self.copy** line can be added.

**Reply by nlohmann** [reply]
@nic-godz Do you need further assistance with this issue?
